File: C#/TheOpenMathLibrary.Engineering/FluidDynamics.cs
namespace TheOpenMathLibrary.Engineering
{
    /// <summary>
    /// Provides helper formulas for introductory fluid-dynamics calculations.
    /// </summary>
    public class FluidDynamics
    {
        /// <summary>
        /// Calculates average flow velocity in a circular pipe from volumetric flow rate.
        /// </summary>
        /// <param name="flowRate">The volumetric flow rate.</param>
        /// <param name="pipeDiameter">The pipe diameter.</param>
        /// <returns>The average flow velocity.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="pipeDiameter"/> is zero.</exception>
        public static double FlowVelocity(double flowRate, double pipeDiameter)
        {
            EnsureNonZero(pipeDiameter, nameof(pipeDiameter));
            return 4d * flowRate / (Math.PI * pipeDiameter * pipeDiameter);
        }

        /// <summary>
        /// Calculates a velocity pseudovector magnitude from linear velocity and radius.
        /// </summary>
        /// <param name="velocity">The velocity.</param>
        /// <param name="radius">The radius.</param>
        /// <returns>The pseudovector magnitude.</returns>
        public static double VelocityPseudovector(double velocity, double radius)
        {
            return velocity * radius;
        }

        /// <summary>
        /// Calculates volume flux through a circular cross section.
        /// </summary>
        /// <param name="flowRate">The volumetric flow rate.</param>
        /// <param name="radius">The radius of the cross section.</param>
        /// <returns>The volume flux.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="radius"/> is zero.</exception>
        public static double VolumeFlux(double flowRate, double radius)
        {
            EnsureNonZero(radius, nameof(radius));
            return flowRate / (Math.PI * radius * radius);
        }

        /// <summary>
        /// Calculates mass current per unit volume.
        /// </summary>
        /// <param name="massCurrent">The mass current.</param>
        /// <param name="volume">The volume.</param>
        /// <returns>The mass current per unit volume.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="volume"/> is zero.</exception>
        public static double MassCurrentPerVolume(double massCurrent, double volume)
        {
            EnsureNonZero(volume, nameof(volume));
            return massCurrent / volume;
        }

        /// <summary>
        /// Calculates mass flow rate from density and volume flux.
        /// </summary>
        /// <param name="density">The fluid density.</param>
        /// <param name="volumeFlux">The volume flux.</param>
        /// <returns>The mass flow rate.</returns>
        public static double MassFlowRate(double density, double volumeFlux)
        {
            return density * volumeFlux;
        }

        /// <summary>
        /// Calculates mass current density.
        /// </summary>
        /// <param name="density">The fluid density.</param>
        /// <param name="velocity">The fluid velocity.</param>
        /// <returns>The mass current density.</returns>
        public static double MassCurrentDensity(double density, double velocity)
        {
            return density * velocity;
        }

        /// <summary>
        /// Calculates momentum current density.
        /// </summary>
        /// <param name="density">The fluid density.</param>
        /// <param name="velocity">The fluid velocity.</param>
        /// <returns>The momentum current density.</returns>
        public static double MomentumCurrentDensity(double density, double velocity)
        {
            return density * velocity * velocity;
        }

        /// <summary>
        /// Calculates pressure gradient magnitude over a distance.
        /// </summary>
        /// <param name="pressure">The pressure difference.</param>
        /// <param name="distance">The distance over which the pressure changes.</param>
        /// <returns>The pressure gradient.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="distance"/> is zero.</exception>
        public static double PressureGradient(double pressure, double distance)
        {
            EnsureNonZero(distance, nameof(distance));
            return pressure / distance;
        }

        /// <summary>
        /// Calculates buoyancy force.
        /// </summary>
        /// <param name="density">The fluid density.</param>
        /// <param name="volume">The displaced volume.</param>
        /// <param name="gravity">The gravitational acceleration.</param>
        /// <returns>The buoyancy force.</returns>
        public static double BuoyancyForce(double density, double volume, double gravity)
        {
            return density * volume * gravity;
        }

        /// <summary>
        /// Evaluates a Bernoulli-style energy-per-volume expression.
        /// </summary>
        /// <param name="pressure">The static pressure term.</param>
        /// <param name="density">The fluid density.</param>
        /// <param name="velocity">The fluid velocity.</param>
        /// <param name="height">The height term.</param>
        /// <param name="gravity">The gravitational acceleration.</param>
        /// <returns>The Bernoulli expression value.</returns>
        public static double BernoullisEquation(double pressure, double density, double velocity, double height, double gravity)
        {
            return pressure + density * velocity * velocity + density * gravity * height;
        }

        /// <summary>
        /// Evaluates an Euler-equation-style energy expression.
        /// </summary>
        /// <param name="pressure">The pressure term.</param>
        /// <param name="density">The fluid density.</param>
        /// <param name="velocity">The fluid velocity.</param>
        /// <param name="height">The height term.</param>
        /// <param name="gravity">The gravitational acceleration.</param>
        /// <returns>The expression value.</returns>
        public static double EulersEquations(double pressure, double density, double velocity, double height, double gravity)
        {
            return pressure + density * velocity * velocity + density * gravity * height;
        }

        /// <summary>
        /// Calculates convective acceleration from velocity and acceleration scale terms.
        /// </summary>
        /// <param name="velocity">The fluid velocity.</param>
        /// <param name="acceleration">The acceleration scale.</param>
        /// <returns>The convective acceleration estimate.</returns>
        public static double ConvectiveAcceleration(double velocity, double acceleration)
        {
            return velocity * acceleration;
        }

        /// <summary>
        /// Evaluates a Navier-Stokes-style scalar expression.
        /// </summary>
        /// <param name="pressure">The pressure term.</param>
        /// <param name="density">The fluid density.</param>
        /// <param name="velocity">The fluid velocity.</param>
        /// <param name="height">The height term.</param>
        /// <param name="gravity">The gravitational acceleration.</param>
        /// <returns>The expression value.</returns>
        public static double NavierStokesEquations(double pressure, double density, double velocity, double height, double gravity)
        {
            return pressure + density * velocity * velocity + density * gravity * height;
        }

        private static void EnsureNonZero(double value, string parameterName)
        {
            if (value == 0d)
            {
                throw new ArgumentOutOfRangeException(parameterName, "The value must not be zero.");
            }
        }
    }
}

File: Rust/src/engineering/fluid_dynamics.rs
//! Fluid-dynamics formulas for flow, pressure, and dimensionless quantities.

/// Provides helper formulas for introductory fluid-dynamics calculations.
pub struct FluidDynamics;

impl FluidDynamics {
    /// Computes mean flow velocity in a circular pipe from volumetric flow rate and diameter.
    pub fn flow_velocity(flow_rate: f64, pipe_diameter: f64) -> f64 {
        (4.0 * flow_rate) / (std::f64::consts::PI * pipe_diameter * pipe_diameter)
    }

    /// Computes tangential speed from angular velocity and radius.
    pub fn velocity_pseudovector(angular_velocity: f64, radius: f64) -> f64 {
        angular_velocity * radius
    }

    /// Computes volumetric flux through a circular cross-section.
    pub fn volume_flux(velocity: f64, radius: f64) -> f64 {
        velocity * std::f64::consts::PI * radius * radius
    }

    /// Computes mass current per unit volume.
    pub fn mass_current_per_volume(mass_current: f64, volume: f64) -> f64 {
        mass_current / volume
    }

    /// Computes mass flow rate.
    pub fn mass_flow_rate(density: f64, volume_flux: f64) -> f64 {
        density * volume_flux
    }

    /// Computes mass current density.
    pub fn mass_current_density(density: f64, velocity: f64) -> f64 {
        density * velocity
    }

    /// Computes momentum current density.
    pub fn momentum_current_density(density: f64, velocity: f64) -> f64 {
        density * velocity * velocity
    }

    /// Computes a pressure gradient.
    pub fn pressure_gradient(pressure_difference: f64, distance: f64) -> f64 {
        pressure_difference / distance
    }

    /// Computes buoyancy force.
    pub fn buoyancy_force(density: f64, volume: f64, gravity: f64) -> f64 {
        density * volume * gravity
    }

    /// Evaluates Bernoulli's equation in pressure form.
    pub fn bernoullis_equation(
        pressure: f64,
        density: f64,
        velocity: f64,
        height: f64,
        gravity: f64,
    ) -> f64 {
        pressure + 0.5 * density * velocity * velocity + density * gravity * height
    }

    /// Evaluates Euler's equation in specific-energy form.
    pub fn eulers_equations(
        pressure: f64,
        density: f64,
        velocity: f64,
        height: f64,
        gravity: f64,
    ) -> f64 {
        pressure / density + 0.5 * velocity * velocity + gravity * height
    }

    /// Computes convective acceleration.
    pub fn convective_acceleration(velocity: f64, velocity_gradient: f64) -> f64 {
        velocity * velocity_gradient
    }

    /// Evaluates a simplified Navier-Stokes acceleration balance.
    pub fn navier_stokes_equations(
        pressure_gradient: f64,
        density: f64,
        body_force: f64,
        velocity_laplacian: f64,
        dynamic_viscosity: f64,
    ) -> f64 {
        -pressure_gradient / density + body_force + dynamic_viscosity * velocity_laplacian / density
    }

    /// Computes the Reynolds number.
    pub fn reynolds_number(
        density: f64,
        velocity: f64,
        characteristic_length: f64,
        dynamic_viscosity: f64,
    ) -> f64 {
        density * velocity * characteristic_length / dynamic_viscosity
    }
}

#[cfg(test)]
mod tests {
    use super::FluidDynamics;

    #[test]
    fn volumetric_flow_relations_are_consistent() {
        let flow_rate = 2.0;
        let radius = 0.5;
        let diameter = 2.0 * radius;
        let velocity = FluidDynamics::flow_velocity(flow_rate, diameter);
        let reconstructed_flow_rate = FluidDynamics::volume_flux(velocity, radius);

        assert!((reconstructed_flow_rate - flow_rate).abs() < 1e-12);
    }

    #[test]
    fn bernoulli_and_reynolds_helpers_match_known_values() {
        let bernoulli = FluidDynamics::bernoullis_equation(1000.0, 2.0, 3.0, 5.0, 9.81);
        let reynolds = FluidDynamics::reynolds_number(1000.0, 2.0, 0.5, 0.001);

        assert!((bernoulli - 1107.1).abs() < 1e-9);
        assert!((reynolds - 1_000_000.0).abs() < 1e-6);
    }
}
